Added org-generate-with-export

hubisan opened the pull request:
This solves #3

Decided to make this a wrapping function and not touch org-generate. Mentioned it in the README.org and added some examples in a separate file.

The function works as follows:

1. The function takes the relevant content as string. This is needed as when exporting all macros are expanded even in parts that are not exported.
2. The strings are concatenated and exported as org. There is an built-in org function which takes a string.
3. The exported string is then inserted into a temp buffer.
4. Org-generate is called with lexical binding the buffer to that temporary buffer.

## No tests added so far

Not familiar with cort, not sure how the tests work atm with setup and teardown.

As it is actually just a wrapper around `org-generate` just checking if the relevant parts get exported with a macro replacement should do it. When callling `(org-generate-with-export "Test/Target")` on this:

```org-mode
#+TITLE: Test
#+AUTHOR: Tester
#+MACRO: test success
* Test
:PROPERTIES:
:CUSTOM_ID: test-value
:END:
#+MACRO: test-1 success
** Not exported
:PROPERTIES:
:org-generate-root: ~/projects/
:END:
** Target
:PROPERTIES:
:org-generate-root: ~/test/
:END:
#+MACRO: test-2 success
{{{test}}}
{{{test-1}}}
{{{test-2}}}
* Another 
** Not exported
```

Should be exported to:

```org-mode
# Created 2020-07-13 Mon 17:41
#+TITLE: Test
#+AUTHOR: Tester
#+macro: test success
* Test
:PROPERTIES:
:CUSTOM_ID: test-value
:END:
#+macro: test-1 success
** Target
:PROPERTIES:
:org-generate-root: ~/test/
:END:
#+macro: test-2 success
success
success
success
```

Could advice `org-generate` and replace to check the content of the temporary buffer before `org-generate` is called.

hubisan committed:
Added org-generate-with-export

Decided to make this a wrapping function and not touch
org-generate.

Mentionned it in the README.org and added some examples in a
separate file.

The function works as follows:

1. The function takes the relevant content as string.
   This is needed as when exporting all macros are expanded even in
   parts that are not exported.
2. The strings are concatenated and exported as org. There is an
   built-in org function which takes a string.
3. The exported string is then inserted into a temp buffer.
4. Org-generate is called with lexical binding the buffer to that
   temporary buffer.

hubisan commented:
Would be better to move the get-string and export-string into separate functions. Would also make testing a lot easier. Will uodate the branch.

hubisan committed:
splitted org-generate-with-export into 3 functions

conao3 commented:
Awesome work!

One question, `org-generate-with-export` could be merge `org-generate`?
It is difficult for the user to fully understand if the tree should be `org-export`ed before it is expanded.
Even if it isn't need `org-export`, I understand that there is no missing information with `org-export`, so it should be reasonable, I think.

conao3 commented:
Please also resolve checkdoc warnings.
```
 org-gen…   206     info            Probably "Returns" should be imperative "Return" (emacs-lisp-checkdoc)
 org-gen…   235     info            Probably "returns" should be imperative "return" (emacs-lisp-checkdoc)
```

Is it not possible to modify the `org-generate` function because there is no suitable test case?
I will add it.

conao3 committed:
define cort--file-contents and use it

conao3 committed:
use simple variable for testing

conao3 committed:
difine dir at toplevel

conao3 committed:
add cort prefix for dir variable

conao3 committed:
define with-cort--org-generate-buffer and use it

conao3 committed:
guess file path is based on cort--dir

conao3 committed:
define cort-deftest--org-generate and use it

conao3 committed:
add heading-with-macro testcase

conao3 committed:
add missing file handling

conao3 committed:
use save-excursion to restore pointer location

conao3 committed:
add empty file name handling

hubisan committed:
fix checkdoc warnings

conao3 committed:
add org-generate-show-save-message variable

conao3 committed:
no message after save file when testing

hubisan committed:
Merged with-export into org-generate and added a defcustom

By setting org-generate-with-export-as-org to t org-generate will
export the target defintion to org before generating.

hubisan commented:
Merged it into `org-generate` and added a `defucstom` to turn the feature on `org-generate-with-export-as-org`. Adapted the readme and the examples accordingly.

Will have a look at `cort` to write a test. As `org-generate--with-export` returns a buffer with the exported string this shouldn't be a big hassle.

conao3 committed:
add testcases

conao3 commented:
Great work!
I've to try write some test cases using `cort`, please reference how to write.

The test could run using `make test` after installing [keg](https://github.com/conao3/keg.el).
If you have any questions, please feel free to ask.

*Set the :org-generate-root: property with a macro* example did not work...
The corresponding test case is `org-genearte/set-variable-with-macro`.

conao3 committed:
use org-generate instead of org-generate-with-export

conao3 committed:
use org-export at default

hubisan commented:
Cool that you added some tests. Don't have access to Emacs atm.

The example is actually incorrect. Should be

```
#+OPTIONS: prop:t
* hugo
#+NAME: hugo-root
: ~/dev/repos/hugo-src/
#+MACRO: hugo-root (eval (concat ":org-generate-root: " (org-sbe "hugo-root") $1))
** page
:PROPERTIES:
{{{hugo-root(content/blog/)}}}
:END:
```

instead of (`: ~/dev/repos/hugo-src/` is on the wrong line)

```org-mode
#+OPTIONS: prop:t
* hugo
#+NAME: hugo-root
#+MACRO: hugo-root (eval (concat ":org-generate-root: " (org-sbe "hugo-root") $1))
: ~/dev/repos/hugo-src/
** page
:PROPERTIES:
{{{hugo-root(content/blog/)}}}
:END:
```

The line at the top of the tests `#+OPTIONS: prop:t` should be removed as it should take care of that by lexical binding  `org-export-with-properties` to `t`.

hubisan committed:
fix example, fixed test for that example, removed prop:t

- fixed the example
- fixed test case for that example
- removed prop:t from all test, the functions takes care of that
(in the examples is it added to be able to preview what they do with
M-x org-org-export-as-org)
- renamed org-genearte to org-generate in tests

hubisan commented:
Changed the example and modified the test. The test still fails, working on it (the two test after as well).

hubisan committed:
Erase buffer as get-buffer-create reuses the buffer if it exists.

hubisan committed:
fixed last test case and added a commentaries in the readme

hubisan commented:
Fixed the last test case. The tests run successfully when run inside Emacs (evaluated the test file and then executed `(cort-run-test)`. 

But when using `make test` they are stopping on the first test where properties need to be exported. With some trial and error I found out that it is stopped/blocked/never ends at (not sure though):

https://github.com/hubisan/org-generate.el/blob/d692868bfd3ae18fa3a7232fe7a7788698dfbf68/org-generate.el#L160

hubisan committed:
changed readme as export-as-org is the default now

hubisan committed:
confirmation needs to be turned off for testing

hubisan commented:
Had to add this of course: `(org-confirm-babel-evaluate nil)`

But the test still doesn't finish with `make test`:

![image](https://user-images.githubusercontent.com/4160043/87645018-31322300-c74d-11ea-9395-988f2e99d8d5.png)

@conao3  any idea how to get around this?

conao3 commented:
I've stuck the same place... Please wait a moment.

conao3 committed:
declare other package variable via defvar

conao3 committed:
use two spaces after period

conao3 commented:
It seems this diff is needed.
Maybe, Since `ignore` obviously has no side effects, let itself could have been omitted.

```diff
declare other package variable via defvar

1 file changed, 1 insertion(+), 1 deletion(-)
org-generate.el | 2 +-

modified   org-generate.el
@@ -151,12 +151,12 @@ not relevant."
             (lambda (e) (buffer-substring-no-properties (car e) (cdr e)))
             regions))))
 
+(defvar org-export-with-properties)
 (defun org-generate--export-string-as-org (string)
   "Export the STRING as org and return the exported string.
 Properties are exported as well."
   (require 'ox-org)
   (let ((org-export-with-properties t))
-    (ignore org-export-with-properties)
     (org-export-string-as string 'org t)))
 
 (defun org-generate--with-export (heading)
```

conao3 committed:
define override function to test silent

hubisan commented:
looks good now 🙂

conao3 committed:
add include-file test

conao3 committed:
set buffer-file-name and default-directory to omit base dir

conao3 committed:
omit cort-dir, use relative path

conao3 committed:
add split-template-files testcase

conao3 committed:
add noweb-reference testcase

conao3 committed:
tagged v1.0.4

conao3 commented:
Looks also good by me.
Thanks for contributing with such a great feature! :tada:
